pkg/reconciler/retry.go

```diff
@@ -0,0 +1,30 @@
+/*
```

Comment:
I think this will be useful to add in pkg, as we might start using this in other repos as well, such as eventing..
Maybe under: https://github.com/knative/pkg/tree/master/reconciler

Can you check whether knative/eventing started doing this retries in their controllers. I think I saw an issue filed sometime back (similar timing as I filed this one)...

Reply:
looks like they did: https://github.com/knative/eventing/issues/2296. I'll set up a PR to add these to https://github.com/knative/pkg/tree/master/reconciler and ping this PR back once that is in (or if I get pushback)

Reply:
created https://github.com/knative/pkg/pull/1008

Reply:
sounds good! one thing that would be nice, if you can, is to add some UT that does at least one retry... not sure how hard it will be, if you need to mock things, etc. But I think it will also be good  so that you get familiar with the reconciler UTs (in case you are not). Maybe borrow something from serving.

Reply:
cool, I'll look at serving to see what they have wrt reconciler UTs and try to borrow something, otherwise if I cant figure it out I'll ask for help.

Reply:
hi @nachocano ,
I spent some time today looking at the reconciler UTs. I feel like I understand a bit better now how they work. However, I cannot figure out a trivial way to create a test that does retries. Here is my understanding, and a half-baked potential way to achieve this kind of tests (though it seems very involved):
- In order to have a UT that performs retries, we would need to have a reconciler RunClientSet mock that can be customized to produce an error when updating a resource.
- The current testing infra relies on a MakeFactory that creates a reconciler using fake clients
- All these fake clients are generated by client-gen and injection-gen and cannot be customized to provide a mock like behavior to cause retries to be needed (afaiu)
- There seems to be some level of mocking in the reconcilers that could be mimicked for the desired behavior (this is my half baked idea):
  - Some of the reconcilers have a `createClientFn` used to create a client and inject a mock in UTs
  - We could provide a similar function to create the `RunClientSet` so that reconciler implementations would use that. But we would need to generate mocks for all the interfaces in `RunClientSet`, perhaps using some well known mocking library, as I could not find anything in the knative code about mocking.

So, my half baked idea sounds extremely involved, maybe you have other ideas as to how to achieve a test for this?

Reply:
OK, no worries about the UTs... 
I think that Table framework had a WithReactors or something, that you can "inject" failures when performing certain operations... But I haven't used it much... 
We can leave that out for the future...

Reply:
@nachocano thanks for the advice. I had not noticed the `WithReactors` in Table, it does seem to work as intended. I added some basic UTs that include 1 retry to each of the modified resources, let me know if these tests look good as is (extending current tests) or if you want me to split off a separate test func just to validate the retries.

Reply:
They look great!! Thanks for adding them!
Can you just add a comment or update the name of the UTs you channged to say that they are also retrying?

Reply:
added "with retry" to UT names